provisioner/aws-manager.js

```diff
@@ -0,0 +1,696 @@
+'use strict';
+
+var Promise = require('promise');
+var lodash = require('lodash');
+var util = require('util');
+var assert = require('assert');
+var debug = require('debug')('aws-provisioner:aws-manager');
+
+
+/**
+ * AWS EC2 state at a specific moment in time
+ */
+function AwsManager(ec2, keyPrefix, pubKey) {
+  assert(ec2);
+  assert(keyPrefix);
+  assert(pubKey);
+  this.ec2 = ec2;
+  this.keyPrefix = keyPrefix;
+  this.pubKey = pubKey;
+  this.__apiState = {};
+  this.__internalState = {};
+  this.__knownKeyPairs = [];
+}
+
+module.exports = AwsManager;
+
+
+/**
+ * Update the state from the AWS API and return a promise
+ * with no resolution value when completed.
+ */
+AwsManager.prototype.update = function() {
+  var that = this;
+
+  var p = Promise.all([
+    that.ec2.describeInstances({
+      Filters: [{
+        Name: 'key-name',
+        Values: [that.keyPrefix + '*']
+      },{
+        Name: 'instance-state-name',
+        Values: ['running', 'pending']
+      }
+    ]}),
```

Comment:
Well, it could be done per-region in multi-region-promised-aws by doing the per-region paging there, but that transforms that library from a dumb wrapper into something with logic and I'm not sure if we want that.